File: 746. Min Cost Climbing Stairs/solution.cpp
/*
392 ms
86.7 MB

TC : O(N)
SC : O(N)
*/

/**
 * @author : SahilK-027
 * @brief : Reciursive solution  + DP 
*/

class Solution {
private:
    int solve(int n, vector<int>& cost, vector<int>& dp){
        // Cannot reach
        if(n < 0 ){
            return 0;
        }
        // Reached out of array or on last index
        if(n == 0 || n == 1){
            return cost[n];
        }
        // If answer is already in DP
        if(dp[n] != -1) return dp[n];
        // Return min answer
        return dp[n] = cost[n] + min(solve(n-1, cost, dp), solve(n-2, cost, dp));
    }
public:
    int minCostClimbingStairs(vector<int>& cost) {
        int n = cost.size();
        vector<int>dp(n+1, -1);
        return min(solve(n-1, cost, dp), solve(n-2, cost, dp));
    }
};

File: 746. Min Cost Climbing Stairs/solution.js
/*
63 ms
42.3 MB

TC: O(n)
SC: O(n)

@author: SahilK-027
@breif: Recursion + Memoization
*/
/**
 * @param {number[]} cost
 * @return {number}
 */

function solve(cost, k, dp){
    // Base case
    if( k == 0 ){
        return cost [0];
    }
    if( k == 1 ){
        return cost[1];
    }
    // If cached then return 
    if(dp[k] != -1){
        return dp[k];
    }
    // Otherwise store ans in dp and then return
    return dp[k] = Math.min(solve(cost, k - 1, dp), solve(cost, k - 2, dp)) + cost[k];
}
var minCostClimbingStairs = function(cost) {
    // Extracting size of input so to get number of stairs
    let n = cost.length;

    // Initializing DP of size n + 1 with -1
    let dp = new Array(n + 1);
    dp.fill(-1);

    // Getting ans as minimum of 1 step call and 2 step call
    let ans = Math.min(solve(cost, n-1, dp), solve(cost, n-2, dp));
    return ans;
};
